eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,595 @@
+	blockEvent.Type = blockwatch.Added
+	info, err = sw.GetSenderInfo(stubSender)
+	require.Nil(err)
+	require.Zero(info.ThawRound.Cmp(startThawR))
+	require.Equal(info.ReserveState, pm.Frozen)
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Equal(info.ReserveState, pm.Frozen)
+	expectedThawRound := new(big.Int).Add(new(big.Int).SetBytes(newReserveFrozenEvent.Data[:32]), big.NewInt(2))
+	assert.Zero(expectedThawRound.Cmp(info.ThawRound))
+
+	// If we don't care about the address, don't handle the event
+	s := pm.RandAddress()
+	sender := ethcommon.LeftPadBytes(s.Bytes(), 32)
+	var senderTopic ethcommon.Hash
+	copy(senderTopic[:], sender[:])
+	newReserveFrozenEvent.Topics[1] = senderTopic
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok = sw.senders[s]
+	assert.False(ok)
+}
+
+func TestUnlockEvent(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	startWithdrawBlock := big.NewInt(5)
+	lpEth := &eth.StubClient{
+		SenderInfo: &pm.SenderInfo{
+			WithdrawBlock: big.NewInt(5),
+		},
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newUnlockEvent := newStubUnlockLog()
+	header.Logs = append(header.Logs, newUnlockEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil don't handle event
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok := sw.senders[stubSender]
+	assert.False(ok)
+
+	// if block removed, make RPC call
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err := sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.WithdrawBlock.Cmp(startWithdrawBlock))
+
+	// If map entry exists parse event log
+	blockEvent.Type = blockwatch.Added
+	info, err = sw.GetSenderInfo(stubSender)
+	require.Nil(err)
+	require.Equal(info.WithdrawBlock, startWithdrawBlock)
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.WithdrawBlock.Cmp(big.NewInt(150)))
+
+	// If we don't care about the address, don't handle the event
+	s := pm.RandAddress()
+	sender := ethcommon.LeftPadBytes(s.Bytes(), 32)
+	var senderTopic ethcommon.Hash
+	copy(senderTopic[:], sender[:])
+	newUnlockEvent.Topics[1] = senderTopic
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok = sw.senders[s]
+	assert.False(ok)
+}
+
+func TestUnlockCancelledEvent(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	lpEth := &eth.StubClient{
+		SenderInfo: &pm.SenderInfo{
+			WithdrawBlock: big.NewInt(0),
```

Comment:
I suggest making this a non-zero value to differentiate between the case where we use an RPC call to refresh the cached `WithdrawBlock` and the case where `WithdrawBlock = 0` because we processed an `UnlockCancelled` event. At the moment, we wouldn't be able to differentiate between the two cases because `WithdrawBlock` is always 0.

Reply:
fixed in 4530eff